fdbserver/DDTeamCollection.actor.cpp

```diff
@@ -5087,3 +5088,606 @@ Future<Void> DDTeamCollection::run(Reference<DDTeamCollection> teamCollection,
+	state int maxTeams = SERVER_KNOBS->MAX_TEAMS_PER_SERVER * processSize;
+	state int teamSize = 3;
+	// state int targetTeamsPerServer = SERVER_KNOBS->DESIRED_TEAMS_PER_SERVER * (teamSize + 1) / 2;
+	state std::unique_ptr<DDTeamCollection> collection = testTeamCollection(teamSize, policy, processSize);
+
+	collection->addTeam(std::set<UID>({ UID(1, 0), UID(2, 0), UID(3, 0) }), true);
+	collection->addTeam(std::set<UID>({ UID(1, 0), UID(3, 0), UID(4, 0) }), true);
+
+	state int result = collection->addTeamsBestOf(8, desiredTeams, maxTeams);
+
+	ASSERT(result >= 8);
+
+	for (auto process = collection->server_info.begin(); process != collection->server_info.end(); process++) {
+		auto teamCount = process->second->teams.size();
+		ASSERT(teamCount >= 1);
+		// ASSERT(teamCount <= targetTeamsPerServer);
+	}
+
+	return Void();
+}
+
+// Due to the randomness in choosing the machine team and the server team from the machine team, it is possible that
+// we may not find the remaining several (e.g., 1 or 2) available teams.
+// It is hard to conclude what is the minimum number of  teams the addTeamsBestOf() should create in this situation.
+TEST_CASE("/DataDistribution/AddTeamsBestOf/NotEnoughServers") {
+	wait(Future<Void>(Void()));
+
+	Reference<IReplicationPolicy> policy =
+	    Reference<IReplicationPolicy>(new PolicyAcross(3, "zoneid", Reference<IReplicationPolicy>(new PolicyOne())));
+	state int processSize = 5;
+	state int desiredTeams = SERVER_KNOBS->DESIRED_TEAMS_PER_SERVER * processSize;
+	state int maxTeams = SERVER_KNOBS->MAX_TEAMS_PER_SERVER * processSize;
+	state int teamSize = 3;
+	state std::unique_ptr<DDTeamCollection> collection = testTeamCollection(teamSize, policy, processSize);
+
+	collection->addTeam(std::set<UID>({ UID(1, 0), UID(2, 0), UID(3, 0) }), true);
+	collection->addTeam(std::set<UID>({ UID(1, 0), UID(3, 0), UID(4, 0) }), true);
+
+	collection->addBestMachineTeams(10);
+	int result = collection->addTeamsBestOf(10, desiredTeams, maxTeams);
+
+	if (collection->machineTeams.size() != 10 || result != 8) {
+		collection->traceAllInfo(true); // Debug message
+	}
+
+	// NOTE: Due to the pure randomness in selecting a machine for a machine team,
+	// we cannot guarantee that all machine teams are created.
+	// When we chnage the selectReplicas function to achieve such guarantee, we can enable the following ASSERT
+	ASSERT(collection->machineTeams.size() == 10); // Should create all machine teams
+
+	// We need to guarantee a server always have at least a team so that the server can participate in data distribution
+	for (auto process = collection->server_info.begin(); process != collection->server_info.end(); process++) {
+		auto teamCount = process->second->teams.size();
+		ASSERT(teamCount >= 1);
+	}
+
+	// If we find all available teams, result will be 8 because we prebuild 2 teams
+	ASSERT(result == 8);
+
+	return Void();
+}
+
+TEST_CASE("/DataDistribution/GetTeam/NewServersNotNeeded") {
+	wait(Future<Void>(Void()));
```

Comment:
Since there is another `wait` below, can this be removed?

Reply:
I haven't looked into this yet - My working assumption was that this has to do with triggering/starting the test.